### PR TITLE
Fix typo in classification function selection logic to improve code consistency

### LINKED ISSUE #32030: Minor typo in ImageClassificationPipeline
### System Info

Issue is on HEAD on the main branch, been around for a while

Transformers env not really relevant, but for completeness:
- `transformers` version: 4.32.1
- Platform: macOS-14.5-arm64-arm-64bit
- Python version: 3.11.5
- Huggingface_hub version: 0.15.1
- Safetensors version: 0.3.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.1.2 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@Narsil There's a typo in ImageClassificationPipeline, that I don't believe usually has any effect, but it's confusing when reading the code. I will provide a PR with a fix.

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

This typo does not cause behavior change, because of the `or self.model.config.num_labels == 1`/`> 1` condition.

### Expected behavior

The `problem_type` condition should match the `num_labels` condition

### PR DESCRIPTION
**Pull request by moses**
Fixes https://github.com/huggingface/transformers/issues/32030 
(previously tagged ~#32013~ )

The existing logic has a typo in it, where, for example, it selects a sigmoid classification function if the problem type is "multi_label_classification" or num_labels is 1. The num_labels condition overrides the problem_type condition, so in practice, I don't think this error will cause a bug, but it's confusing.

This fix corrects the problem type conditions.

Please review at your convenience @Narsil

### CODE REVIEW

**Comment by leloykun** [comment]
Hi! I think you should be tagging this? https://github.com/huggingface/transformers/issues/32030

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!
